app/models/authenticator.rb

```diff
@@ -1,14 +1,25 @@
 module Authenticator
+  require_nested :Base
+  require_nested :Database
+  require_nested :Httpd
+  require_nested :Ldap
+
   def self.for(config, username = nil)
-    subclass_for(config, username).new(config)
+    if username == 'admin'
+      Database.new(config)
+    else
+      authenticator_class(config[:mode]).new(config)
+    end
   end
 
-  def self.subclass_for(config, username)
-    return Database if username == "admin"
-
-    mode_class_name = config[:mode].camelize
-    mode_class_name = 'Ldap' if mode_class_name == 'Ldaps'
+  def self.authenticator_class(name)
+    name = name.to_s
+    Base.subclasses.find do |subclass|
+      return subclass if subclass.authenticates_for.include?(name)
+    end
 
-    "Authenticator::#{mode_class_name}".constantize
+    # try to constantize in case the subclass is not loaded yet, but it still exists
+    # this is happens with authenticators from provider gems
+    "Authenticator::#{name.camelize}".safe_constantize
```

Comment:
see the comment why the above `.find` is not enough.

I could fix that by loading `Authenticator::Amazon` in an initializer in the provider